### ISSUE TITLE
Credential Management Level 1 - Federated Credentials

### ISSUE THREAD

**Issue by elf-pavlik** [solid/solid#142]
It seems to work with OIDC

https://w3c.github.io/webappsec-credential-management/#federated
> protocol, of type DOMString, readonly, nullable
    The credential’s federated identity provider’s protocol (e.g. "openidconnect"). If the value is null, then the protocol can be inferred from the provider.

It also should provide friendly and uniform UX in web browsers
https://w3c.github.io/webappsec-credential-management/#user-mediated-selection
![credential selection](https://w3c.github.io/webappsec-credential-management/mock-chooser.png)

demo: https://credential-management-sample.appspot.com/

I think this look like something to consider for solid to start adopting in a near future.

**Comment by elf-pavlik** [comment]
This one might at some point supersede WebID-TLS 
https://developer.mozilla.org/en-US/docs/Web/API/PublicKeyCredential

**Comment by dmitrizagidulin** [comment]
> This one might at some point supersede WebID-TLS

I'm not sure. The PublicKeyCredential is bound to (and can only be accessed by) a single origin. It's made for a very different purpose.

**Comment by elf-pavlik** [comment]
Good point!

**Comment by RubenVerborgh** [comment]
That looks really lovely actually. @timbl Should the Solid team get involved in https://w3c.github.io/webappsec-credential-management/?

**Comment by elf-pavlik** [comment]
It seems to have a limitation that credentials sharing stays restricted to particular origin (possibly with sub domains) https://w3c.github.io/webappsec-credential-management/#security-credential-access

When WebID-OIDC solid client encounters new origin, it would still require presenting an input for WebID possibly with tiny UX optimization using https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-autocomplete

**Comment by RubenVerborgh** [comment]
> When WebID-OIDC solid client encounters new origin, it would still require presenting an input

And that's how it should be, privacy-wise, but hopefully not with a modal popup.

**Comment by dmitrizagidulin** [comment]
@elf-pavlik & @RubenVerborgh 

> When WebID-OIDC solid client encounters new origin

So, this is an interesting situation, and there's a couple of things to keep in mind. 

One, it all depends on what we mean by "encounters a new origin". In the OIDC family of protocols, the User Consent UI is displayed only when *signing into an RP/App*, for the first time. Not the Resource Server. This makes sense, especially keeping in mind that interrupting a user's flow with a Consent UI is expensive and disruptive, and only useful in times where there's potential danger. And the danger with an app is that, you're giving it permission to act on your behalf (though not as you, it has its own identity). 

So, when we talk about establishing user sessions, we're only describing two things: 1) A session between a user and their IdP (and it's only a session for easy of re-authorizing new apps, so that the user doesn't have to re-enter their credentials). and 2) A session between a user and the RP / Client App. And it's a session in the sense of -- a) the app needs to display which account/persona a user is using, as part of the UI, and b) it needs to be able to re-use their credentials for various requests.

But no session is created (and no cookie is generated) when the RP makes a request to some third Resource Server. So, when I sign into a distributed Twitter type app which fetches messages from a dozen different resource servers, I only have a session with my WebID's IDP, and the app itself. But not with any of those resource servers.

So, no session is generated with those resource servers, and they are unable to act on a user's behalf (or on the app's behalf, for what it's worth). In that respect, it's definitely not worth paying the expense of a separate Consent step for each one.

The other thing to consider of course, with these multiple third party resource servers, is the tension between a user's desire for privacy and anonymity, and the resource owner's desire for knowing who is accessing their data.

Part of the reason that the current auth client behavior is to first try a request without a bearer token, and to re-send the token only on a 401, is to try as much as possible to preserve the anonymity of the requester. (Though this clashes with the resource owner's desire to know who accessed their resource. Not to mention, results in twice the number of requests, in the worst case.)

On the other hand, if those resources are permissioned (non-public), arguing that the Resource Server shouldn't know the WebID of the requester makes less sense. And so, again there's no remote reason for a consent step when encountering a new Resource Server origin.

**Comment by elf-pavlik** [comment]
> But no session is created (and no cookie is generated) when the RP makes a request to some third Resource Server. So, when I sign into a distributed Twitter type app which fetches messages from a dozen different resource servers, I only have a session with my WebID's IDP, and the app itself. But not with any of those resource servers.

If such app runs fully in a web browser (eg. hosted on github pages), I think I only need session with my IdP, very likely also making sure to allow orgin serving that app in ACLs ([related Mozilla Persona consideration](https://developer.mozilla.org/en-US/docs/Archive/Mozilla/Persona/Security_Considerations#Explicitly_specify_the_audience_parameter))

Now twitter with all public content doesn't make best example, if instead we consider distributed issue tracker which uses WAC and each project has tracker in different origin. If my in browser app wants to aggregate issues from 10 different trackers in 10 different origins it has to make *authenticated requests* to each of those origins. With WebID-TLS I would get certificate selection popup for at least on first encounter for those 10 origins. With WebID-OIDC I understand I would have to authenticate with each of those origins to get bearer tokens (or session cookie). The app doesn't really act here as Relying Party and all requests get made directly from the app running in the browser, I think this makes each data origin act as both Resource Server and Relying Party and app just acts as [OAuth 2.0 public client](https://tools.ietf.org/html/rfc6749#section-2.1)

Side tracking a bit, but I wouldn't expect any app origin restrictions in ACLs in those issue trackers, only permissions based on my WebID (just as I wouldn't expect restricted mailing lists to make any assumptions about mail clients used) . For that user in the end would have to make sure onself to only use trusted apps and cookie based sessions might actually expose resources without origin restrictions to get accessed by malicious apps (very likely WebID-TLS as well). Barer tokens managed by trusted apps might actually turn out most resilient. Any suggestions for which repo/issue to discuss those security aspects of ACLs without origin restrictions?

**Comment by dmitrizagidulin** [comment]
> If such app runs fully in a web browser (eg. hosted on github pages), I think I only need session with my IdP

No, you also need a session with that app (specifically, with its auth client). This is important. You need two sessions, and the session with the app is a) so it can display which persona you're using, but more importantly, b) so it can generate a destination-specific Proof of Posession bearer token (in which your ID Token is embedded) for each other Resource Server it contacts.

> The app doesn't really act here as Relying Party and all requests get made directly from the app running in the browser, I think this makes each data origin act as both Resource Server and Relying Party

The JS app (hosted on github or wherever) *is* acting as a Relying Party. (Specifically, as a *Presenter* type of RP). It has the ability to access protected profile resources, it can refresh the bearer token with the IDP, etc. The other Resource Servers on third-party origins have much more limited ability than the traditional Relying Party (they cannot refresh any tokens or anything like that). And no session is made with them (and they do not generate a browser cookie).

> Any suggestions for which repo/issue to discuss those security aspects of ACLs without origin restrictions?

I think https://github.com/solid/web-access-control-spec works for that.

**Comment by dmitrizagidulin** [comment]
> With WebID-OIDC I understand I would have to authenticate with each of those origins to get bearer tokens (or session cookie).

Again, just to clarify (I realize there is a lot of moving parts here), that's not how the mechanism work. The app (or the user) does *not* authenticate with each of those origins, it does not get a bearer token from them, nor a cookie.

**Comment by elf-pavlik** [comment]
thanks @dmitrizagidulin! could you just point me to explanation to how those 10 different origins authenticate requests made by the app? I understood this part of your reply hints it
> so it can generate a destination-specific Proof of Posession bearer token (in which your ID Token is embedded) for each other Resource Server it contacts.

**Comment by dmitrizagidulin** [comment]
Oof, yeah so, I'm a bit behind on the explanation part, still working on a PR to the OIDC spec. So.. soon :)
Meanwhile, the best I can offer is the ol' look through the source code -- this may be a good starting place https://github.com/anvilresearch/oidc-rs/blob/master/src/PoPToken.js (plus there's links to the two RFCs on which it's based). 

But yes, writeup coming soon.
